File: graphs/kruskal.cc
// https://www.hackerrank.com/challenges/kruskalmstrsub/problem
#include <bits/stdc++.h>
using namespace std;

struct Edge : public pair<int,int>{
  int cost;
};

bool edge_compare(Edge a, Edge b){
  if(a.cost == b.cost)
    return (a.first + a.second < b.first + b.second);
  else
    return a.cost < b.cost;
}

class DisjointSet{
private:
  vector<int> parents;
  vector<int> ranks;
  
public:  
  DisjointSet(int n){
    // Initialize dummy first position
    parents.push_back(-1);
    ranks.push_back(0);
    
    for(int i = 1; i <= n; i++){
      parents.push_back(i);
      ranks.push_back(0);
    }
  }
  
  int find(int node){
    if(node != parents[node]){
      parents[node] = find(parents[node]);
    }
    
    return parents[node];
  }
  
  
  void merge(int first_forest, int second_forest){
    int first_parent = find(first_forest);
    int second_parent = find(second_forest);
    
    if(ranks[first_parent] < ranks[second_parent])
      parents[first_parent] = second_parent;
    else{
      parents[second_parent] = first_parent;
      if(ranks[first_parent] == ranks[second_parent])
        ranks[second_parent]++;
    }
  }
};

int mst(int n, vector<Edge> edges){
  // Order edges with increasing order of cost and sum
  // of cordinates in case of tie
  sort(edges.begin(), edges.end(), edge_compare);
  int result = 0;
  int first_forest, second_forest;
  DisjointSet forests(n);
  
  for(Edge e : edges){
    first_forest = forests.find(e.first);
    second_forest = forests.find(e.second);
    
    if(first_forest != second_forest){
      result += e.cost;
      forests.merge(first_forest, second_forest);
    }
  }
  
  return result;
}

int main() {
  int n;
  int m;
  Edge current;
  cin >> n >> m;
  vector<Edge> edges(m);
  
  for(int edges_i = 0; edges_i < m; edges_i++){
    cin >> current.first >> current.second;
    cin >> current.cost;
    edges[edges_i] = current;
  }
  
  int result = mst(n, edges);
  cout << result << endl;
  return 0;
}

File: graphs/kruskal.py
#!/usr/bin/python3
from operator import attrgetter

class Edge:
    def __init__(self, first, second, cost):
        self.first  = first
        self.second = second
        self.cost   = cost

class DisjointSet:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.rank   = [0 for i in range(n)]

    def find(self, node):
        if node != self.parent[node]:
            self.parent[node] = self.find(self.parent[node])

        return self.parent[node];


    def merge(self, u, v):
        u_parent = self.find(u)
        v_parent = self.find(v)

        if self.rank[u_parent] < self.rank[v_parent]:
            self.parent[u_parent] = v_parent
        else:
            self.parent[v_parent] = u_parent

        if self.rank[u_parent] == self.rank[v_parent]:
            self.rank[v_parent] += 1


def mst(n, edges):
    edges = sorted(edges, key = attrgetter("cost"))
    result = 0
    forests = DisjointSet(n);

    for e in edges:
        first  = forests.find(e.first)
        second = forests.find(e.second)

        if first != second:
            result += e.cost
            forests.merge(first, second)

    return result


# Number of nodes
n = int(input())
# Number of edges
m = int(input())
edges = []

for i in range(m):
    u,v,w = map(int, input().split())
    current = Edge(u,v,w)
    edges.append(current)

result = mst(n, edges)
print(result)
